src/include/sof/wait.h

```diff
@@ -157,4 +160,17 @@ static inline void wait_delay(uint64_t number_of_clks)
 		idelay(PLATFORM_DEFAULT_DELAY);
 }
 
+static inline int wait_for_completion_udelay(int (*condition)(void *data), void *data, int us)
```

Comment:
why not pass the completion_t here instead of a callback, makes it simpler

Reply:
because in some case, there is no completion, just like waiting for a register idle

Reply:
please refer to my second commit in this PR

Reply:
waiting for a register is called polling, and this is probably better named poll_for_condition_timeout()

Reply:
I just want to make a generic solution.  Ok, I will refine my patch